src/Microsoft.ML.Data/Commands/TrainCommand.cs

```diff
@@ -252,77 +252,27 @@ public static string MatchNameOrDefaultOrNull(IExceptionContext ectx, ISchema sc
             ch.CheckValueOrNull(validData);
             ch.CheckValueOrNull(inpPredictor);
 
-            var trainerRmd = trainer as ITrainer<RoleMappedData>;
-            if (trainerRmd == null)
-                throw ch.ExceptUserArg(nameof(TrainCommand.Arguments.Trainer), "Trainer '{0}' does not accept known training data type", name);
-
-            Action<IChannel, ITrainer, Action<object>, object, object, object> trainCoreAction = TrainCore;
-            IPredictor predictor;
             AddCacheIfWanted(env, ch, trainer, ref data, cacheData);
             ch.Trace("Training");
             if (validData != null)
                 AddCacheIfWanted(env, ch, trainer, ref validData, cacheData);
 
-            var genericExam = trainCoreAction.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(
-                typeof(RoleMappedData),
-                inpPredictor != null ? inpPredictor.GetType() : typeof(IPredictor));
-            Action<RoleMappedData> trainExam = trainerRmd.Train;
-            genericExam.Invoke(null, new object[] { ch, trainerRmd, trainExam, data, validData, inpPredictor });
-
-            ch.Trace("Constructing predictor");
-            predictor = trainerRmd.CreatePredictor();
+            var trainerEx = trainer as ITrainerEx;
+            if (inpPredictor != null && trainerEx?.SupportsIncrementalTraining != true)
+            {
+                ch.Warning("Ignoring " + nameof(TrainCommand.Arguments.InputModelFile) +
+                    ": Trainer does not support incremental training.");
+                inpPredictor = null;
+            }
+            ch.Assert(validData == null || CanUseValidationData(trainer));
+            var predictor = trainer.Train(new TrainContext(data, validData, inpPredictor));
```

Comment:
(nit) is `inpPredictor` a type-o?  Should it just be `inPredictor`?

Reply:
I think the intent from the caller was that this be `inputPredictor` so maybe they were just being weird about abbreviating it? I'll change it to `inputPredictor` since `inp` as an abbreviation for `input` seems odd.

Reply:
I could have sworn I edited it, yet I see the file hasn't been changed... should be in next commit.